### DISCUSSION TITLE
GitUI asks for username each time when pulling from remote repository

### DISCUSSION THREAD

**Discussion by jgoldhammer** [extrawurst/gitui#1085] [Q&A]
**Describe the bug**
I want to use gitui to pull and push from a remote repository. Gitui asks for username and password each time.

**To Reproduce**
1. Start gitui in a directory with a cloned repository
2. Start pulling from remote repository

**Expected behavior**
The pull works without entering my username and passwords

**Screenshots**
![image](https://user-images.githubusercontent.com/3872101/150506708-e6009c37-4139-4028-bde7-a683fd0617ed.png)

**Context (please complete the following information):**
 - OS/Distro + Version: macos 12.1 (21C52)
 -  GitUI Version ➜ gitui -V gitui 0.19.0 
 - Rust version: 1.57.0

**Additional context**
Using git pull from commandline works without entering credentials

![image](https://user-images.githubusercontent.com/3872101/150507081-e5163116-95cd-45cb-bc5b-80c0ebdb4768.png)

GitUi Log:

![image](https://user-images.githubusercontent.com/3872101/150506981-16260f7c-5b4f-483b-8ee7-74dc3ed7b91a.png)

**Comment by extrawurst** [comment]
setup an ssh key that will solve it. I don't want gitui in the business of securely storing passwords and ssh is a best practice anyway: https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent

**Reply by extrawurst** [reply]
or am I missing something here? how is it supposed to know your credentials otherwise?

**Reply by jgoldhammer** [reply]
Due to several reasons we are mainly using https on our gitlab instance and do not support ssh protocol.

Git uses a credential manager in the background to support these scenarios and the credentials are stored differently on each platform- on macOS this is the keychain.

More Infos are available here:

https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage

https://github.com/GitCredentialManager/git-credential-manager

**Reply by extrawurst** [reply]
can you do me a favour and run `printf "protocol=https\nhost=github.com\n" | git credential fill` and see if this outputs the correct credentials for you? I am working on integrating this as of #800

**Reply by jgoldhammer** [reply]
Yes, it does @extrawurst ! Thanks for pushing this topic. This will make it more easily to use gitui for me !

**Reply by extrawurst** [reply]
Great! If you could follow the steps to build from the branch to test this I can even make it part of the upcoming 0.20 release: https://github.com/extrawurst/gitui/issues/800#issuecomment-1019382367
